apps/shared/y4m.c

```diff
@@ -201,6 +202,76 @@ static int y4mReadLine(FILE * inputFile, avifRWData * raw, const char * displayF
     return -1;
 }
 
+// Limits each sample value to fit into avif->depth bits and avif->yuvRange.
+// Returns AVIF_TRUE if any sample was clamped this way.
+static avifBool y4mClampSamples(avifImage * avif)
+{
+    avifBool samplesWereClamped = AVIF_FALSE;
+    if (avifImageUsesU16(avif)) {
+        const uint16_t maxSampleValue = (uint16_t)((1u << avif->depth) - 1u);
+        uint16_t minSampleValues[4] = { 0, 0, 0, 0 };
+        uint16_t maxSampleValues[4] = { maxSampleValue, maxSampleValue, maxSampleValue, maxSampleValue };
+        if (avif->yuvRange == AVIF_RANGE_LIMITED) {
+            minSampleValues[AVIF_CHAN_Y] = 16 << (avif->depth - 8);
+            minSampleValues[AVIF_CHAN_U] = 16 << (avif->depth - 8);
+            minSampleValues[AVIF_CHAN_V] = 16 << (avif->depth - 8);
+            maxSampleValues[AVIF_CHAN_Y] = 235 << (avif->depth - 8);
+            maxSampleValues[AVIF_CHAN_U] = 240 << (avif->depth - 8);
+            maxSampleValues[AVIF_CHAN_V] = 240 << (avif->depth - 8);
```

Comment:
We need to use different values for MatrixCoefficients 0 and 8. See H.273, Equations 20-22, 41-43, and 44-46. I think Y, U, V values have the same range for MatrixCoefficients 0 and 8. It would be good to work out the details for MatrixCoefficients 8.